### ISSUE TITLE
Websocket error "invalid frame header" in browser when compression is enabled on a big(ish) payload

### ISSUE THREAD

**Issue by nicoroy2561** [lesismal/nbio#444]
Hi,
thank  you as always for the hard work.

I have come across this scenario (golang nbio server > js browser client):
- I have a relatively bigger json payload than the one I usually transmit (around >100,000 bytes, or 100KB+).
- I have compression enabled but on default level for the connection
- the following happens:
1. the string gets compressed down to ~33-35KB and (seemingly) sent
2. on the browser I get a websocket "invalid frame header" error
3. the connection is terminated by the browser

I have tried:
- disabling compression, which results in the payload being sent without issues
- ensuring the string passed is valid UTF8, which it is 
- setting the compression to higher levels, which results in the payload being sent only if the compression level is high enough that it gets reduced to less than 32KB
- increasing upgrader.MessageLengthLimit, which probably is not meant to have any effect on server-sent data
- setting DefaultMaxWebsocketFramePayloadSize to a value bigger than the default 32KB, which oddly enough didn't matter (tried 128KB)

Following is the code I use:

```
...
	engine = nbhttp.NewEngine(nbhttp.Config{
		Network: "tcp",
		Addrs:   []string{":" + port},
		Handler: mux,

		IOMod:             nbhttp.IOModMixed,
		MaxLoad:           65_000,
		MaxBlockingOnline: 10_000,
	})
...
	engine.Start()
...
```

```
... (upgrade request)
	u := websocket.NewUpgrader()
	u.KeepaliveTime = time.Second * 60 * 30 //30m
	u.EnableCompression(true)
	u.MessageLengthLimit = 1024 * 1024 * 10 // 10MB
...
```

The following, which is supposed to return an error in case one occurs, returns nil:
```
	c.WriteMessage(websocket.TextMessage, m)
```

Please let me know if you need more information, I can try to replicate it and inspect further.

EDIT: forgot to mention, I'm using nbio v1.5.10 (latest) with llib 1.1.13

**Comment by lesismal** [comment]
Thank you for the feedback, work on it.

**Comment by lesismal** [comment]
Please try setting this first:
```golang
upgrader.Engine.MaxWebsocketFramePayloadSize = 1024 * 128(make sure the size is bigger than your data size), 
```
Then it won't send multi frames for this single big message payload and seems working fine.
Because in the IOModMixed, the conn is not handled by the engine's poller and has a default engine inited with 32k, so we need to set it by this way.


By using your example, without set the MaxWebsocketFramePayloadSize to bigger size:
1. I tried gorilla client, it works fine with nbio server.
2. I tried gorilla server, it works fine with browser.

So, nbio server still has the problem, I am still working on it.

**Comment by lesismal** [comment]
Change it to send single frame for compressed message instead of multi frames, it works in my test, please try this commit:
```sh
 go get -u github.com/lesismal/nbio@de9f897a1672f96762c1adbd793a16fbb2c8a868
```

**Comment by nicoroy2561** [comment]
Thank you for the fast responses and handling.

> Then it won't send multi frames for this single big message payload and seems working fine. Because in the IOModMixed, the conn is not handled by the engine's poller and has a default engine inited with 32k, so we need to set it by this way.
Got it. I assume there may be some other limitations, otherwise I'd perhaps suggest carrying over any custom config from the main engine to that used in IOModMixed.



> Change it to send single frame for compressed message instead of multi frames, it works in my test, please try this commit:
> 
> ```shell
>  go get -u github.com/lesismal/nbio@de9f897a1672f96762c1adbd793a16fbb2c8a868
> ```
Trying this now, will come back with feedback.

A quick question though: it is my understanding that, while websocket technically allows very big single frame messages to be sent, browsers usually limit single frames to 1-2MB max. Of course we're talking about compressed values here so it'd be relatively hard to reach those, but could it not be a problem^

**Comment by nicoroy2561** [comment]
> Change it to send single frame for compressed message instead of multi frames, it works in my test, please try this commit:
> 
> ```shell
>  go get -u github.com/lesismal/nbio@de9f897a1672f96762c1adbd793a16fbb2c8a868
> ```
I can confirm this does result in the message being delivered :)
Leaving the issue open for the other questions in my previous answer, but otherwise this can be closed.

**Comment by lesismal** [comment]
Now clear with the reason: the compress flag bit should be set only in the first frame, while the previous implementation set it in all the frames.

I have fixed it and reverted to the multi frames, please try:
```sh
go get -u github.com/lesismal/nbio@afef10f44834679db278264093cd858d87de9550
```

**Comment by nicoroy2561** [comment]
I can confirm this is now fixed.
Thanks a lot for the fast handling :) Love this project!

**Comment by lesismal** [comment]
Thanks Bro! :)
Appreciate your support always!

**Comment by lesismal** [comment]
published: [v1.5.11](https://github.com/lesismal/nbio/releases/tag/v1.5.11)
